File: 5-longest-palindromic-substring/solution.ts
function longestPalindrome(s: string): string {
  if (!s) {
    return "";
  }

  const n = s.length;
  const dp: boolean[][] = new Array(n).fill(0).map(() => new Array(n).fill(false));
  let longest = s[0];

  for (let i = 0; i < n; i++) {
    dp[i][i] = true;
  }

  let maxLen = 1;

  for (let start = n - 1; start >= 0; start--) {
    for (let end = start + 1; end < n; end++) {
      if (s[start] === s[end]) {
        if (end - start === 1 || dp[start + 1][end - 1]) {
          dp[start][end] = true;
          if (end - start + 1 > maxLen) {
            maxLen = end - start + 1;
            longest = s.substring(start, end + 1);
          }
        }
      }
    }
  }

  return longest;
}

File: 5-longest-palindromic-substring/solution.py
def longest_palindromic_substring(s):
    if not s:
        return ""

    n = len(s)
    dp = [[False] * n for _ in range(n)]
    longest = ""

    for i in range(n):
        dp[i][i] = True
        longest = s[i]

    max_length = 1

    for start in range(n - 1, -1, -1):
        for end in range(start + 1, n):
            if s[start] == s[end]:
                if end - start == 1 or dp[start + 1][end - 1]:
                    dp[start][end] = True
                    if end - start + 1 > max_length:
                        max_length = end - start + 1
                        longest = s[start:end + 1]

    return longest

# Example usage:
s = "babad"
result = longest_palindromic_substring(s)
print(result)  # Output: "bab" or "aba"
